domain/application/state/charm.go

```diff
@@ -374,24 +374,65 @@ WHERE uuid = $charmID.uuid;
 
 	stmt, err := s.Prepare(query, archivePath, ident)
 	if err != nil {
-		return "", internalerrors.Errorf("failed to prepare query: %w", err)
+		return "", internalerrors.Errorf("preparing query: %w", err)
 	}
 
 	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		if err := tx.Query(ctx, stmt, ident).Get(&archivePath); err != nil {
 			if errors.Is(err, sqlair.ErrNoRows) {
 				return applicationerrors.CharmNotFound
 			}
-			return internalerrors.Errorf("failed to get charm archive path: %w", err)
+			return err
 		}
 		return nil
 	}); err != nil {
-		return "", internalerrors.Errorf("failed to get charm archive path: %w", domain.CoerceError(err))
+		return "", internalerrors.Errorf("getting charm archive path: %w", domain.CoerceError(err))
 	}
 
 	return archivePath.ArchivePath, nil
 }
 
+// GetCharmArchiveMetadata returns the archive storage path and hash for the
+// charm using the charm ID.
+// If the charm does not exist, a [errors.CharmNotFound] error is returned.
+func (s *State) GetCharmArchiveMetadata(ctx context.Context, id corecharm.ID) (archivePath string, hash string, err error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", "", internalerrors.Capture(err)
+	}
+
+	var archivePathAndHash charmArchivePathAndHash
+	ident := charmID{UUID: id.String()}
+
+	query := `
+SELECT &charmArchivePathAndHash.*
+FROM charm
+JOIN charm_hash ON charm.uuid = charm_hash.charm_uuid
+JOIN hash_kind ON charm_hash.hash_kind_id = hash_kind.id
+WHERE charm.uuid = $charmID.uuid
+AND hash_kind.name = "sha256";
+`
+
+	stmt, err := s.Prepare(query, archivePathAndHash, ident)
+	if err != nil {
+		return "", "", internalerrors.Errorf("preparing query: %w", err)
+	}
+
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := tx.Query(ctx, stmt, ident).Get(&archivePathAndHash); err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return applicationerrors.CharmNotFound
+			}
+			return err
+		}
+		return nil
+	}); err != nil {
+		return "", "", internalerrors.Errorf("getting charm archive metadata: %w", domain.CoerceError(err))
```

Comment:
I don't believe you need to use domain.CoerceError anymore, since its automatically applied to the error result from a Txn